src/resources/rmt_esp32.cc

```diff
@@ -65,33 +84,77 @@ PRIMITIVE(init) {
   return proxy;
 }
 
-PRIMITIVE(use) {
-  ARGS(RMTResourceGroup, resource_group, int, channel_num)
+PRIMITIVE(channel_new) {
+  ARGS(RMTResourceGroup, resource_group, int, memory_block_count, int, channel_num)
   ByteArray* proxy = process->object_heap()->allocate_proxy();
   if (proxy == null) ALLOCATION_FAILED;
+  if (memory_block_count <= 0) INVALID_ARGUMENT;
 
-  rmt_channel_t requested = static_cast<rmt_channel_t>(channel_num);
-  if (!rmt_channels.take(requested)) ALREADY_IN_USE;
+  rmt_channel_t channel = kInvalidChannel;
 
-  IntResource* resource = resource_group->register_id(channel_num);
-  if (!resource) {
-    rmt_channels.put(requested);
-    MALLOC_FAILED;
+  if (channel_num == -1 && memory_block_count == 1) {
+    channel = rmt_channels.any();
+  } else if (memory_block_count == 1) {
+    channel = static_cast<rmt_channel_t>(channel_num);
+    if (!rmt_channels.take(channel)) channel = kInvalidChannel;
+  } else {
+    // Try to find adjacent channels that are still free.
+    int current_start_id = channel_num == -1 ? 0 : channel_num;
+    while (current_start_id + memory_block_count <= SOC_RMT_CHANNELS_NUM) {
+      int taken = 0;
+      for (int i = 0; i < memory_block_count; i++) {
+        bool succeeded = rmt_channels.take(static_cast<rmt_channel_t>(current_start_id + i));
+        if (!succeeded) break;
+        taken++;
+      }
+      if (taken == memory_block_count) {
+        // Success. We have reserved channels that are next to each other.
+        channel = static_cast<rmt_channel_t>(current_start_id);
+        break;
+      } else {
+        // Release all the channels we have reserved, and then try at a later
+        // position.
+        for (int i = 0; i < taken; i++) {
+          rmt_channels.put(static_cast<rmt_channel_t>(current_start_id + i));
+        }
+        if (channel_num == -1) {
+          // Continue searching after the current failure.
+          current_start_id += taken + 1;
+        } else {
+          // Failure. Couldn't allocate the requested memory blocks at this position.
+          break;
+        }
+      }
+    }
   }
+  if (channel == kInvalidChannel) ALREADY_IN_USE;
+
+  RMTResource* resource = null;
+  { HeapTagScope scope(ITERATE_CUSTOM_TAGS + EXTERNAL_BYTE_ARRAY_MALLOC_TAG);
+    resource = _new RMTResource(resource_group, channel, memory_block_count);
+    if (!resource) {
+      for (int i = 0; i < memory_block_count; i++) {
```

Comment:
Can this code be shared with the same code in the on_unregister_resource method?

Reply:
I think it's ok to leave the loop here.
It's a typical pattern: whatever was taken in the primitive must be returned if something goes wrong. I think it would be harder to ensure this if we called out to somewhere else.